Documentation/driver-api/ancillary_bus.rst

```diff
@@ -0,0 +1,218 @@
+shared object(s) will be allocated by the registering driver.  This layout
+allows the ancillary_driver's registering module to perform a container_of()
+call to go from the pointer to the ancillary_device, that is passed during the
+call to the ancillary_driver's probe function, up to the parent object, and then
+have access to the shared object(s).
+
+The memory for the ancillary_device will be freed only in its release()
+callback flow as defined by its registering driver.
+
+The memory for the shared object(s) must have a lifespan equal to, or greater
+than, the lifespan of the memory for the ancillary_device.  The ancillary_driver
+should only consider that this shared object is valid as long as the
+ancillary_device is still registered on the ancillary bus.  It is up to the
+registering driver to manage (e.g. free or keep available) the memory for the
+shared object beyond the life of the ancillary_device.
+
+The ancillary_device cannot have a lifespan longer than the lifespan of the
+registering driver.
+
+Ancillary Drivers
+=================
+
+Ancillary drivers follow the standard driver model convention, where
+discovery/enumeration is handled by the core, and drivers
+provide probe() and remove() methods. They support power management
+and shutdown notifications using the standard conventions.
+
+.. code-block:: c
+
+	struct ancillary_driver {
+		int (*probe)(struct ancillary_device *);
+		int (*remove)(struct ancillary_device *);
+		void (*shutdown)(struct ancillary_device *);
+		int (*suspend)(struct ancillary_device *, pm_message_t);
+		int (*resume)(struct ancillary_device *);
+		struct device_driver driver;
+		const struct ancillary_device_id *id_table;
+	};
+
+Ancillary drivers register themselves with the bus by calling
+ancillary_register_driver(). The id_table contains the names of ancillary
+devices that a driver can bind with.
+
+Example Usage
+=============
+
+Ancillary devices are created and registered by a subsystem-level core device
+that needs to break up its functionality into smaller fragments. One way to
+extend the scope of an ancillary_device would be to encapsulate it within a
+domain-specific structure defined by the parent device. This structure contains
+the ancillary_device and any associated shared data/callbacks needed to
+establish the connection with the parent.
+
+An example would be:
+
+.. code-block:: c
+
+        struct foo {
+		struct ancillary_device adev;
+		void (*connect)(struct ancillary_device *adev);
+		void (*disconnect)(struct ancillary_device *adev);
+		void *data;
+        };
+
+The parent device would then register the ancillary_device by calling
+ancillary_register_device() with the pointer to the adev member of the above
+structure. The parent would provide a match_name for the ancillary_device that
+will be used for matching and binding with a driver.
+
+For the binding to succeed when an ancillary_device is registered, there needs
```

Comment:
this is not true.

ancillary driver can be loaded several hours after the ancillary device is created.
So it is not_when_ancillary_device_is_registered.
binding can happen later.

Just say,
Whenever an ancillary_driver is registered, based on the match_name provided, driver's probe() is invoked for the matching devices.

Reply:
changed